server: trim deps in StartTenant

tbg opened the pull request:
`(*TestServer).StartTenant` now uses only the Gossip instance(*) from
the TestServer and sets everything else up from scratch. This
demonstrates that mod the Gossip dependency, we're en route to
running the SQL server in a standalone process.

Informs #49104.

(*): and the Stopper, but that's fine.

Release note: None

tbg committed:
cli: ensure SQL prefix for SQL-specific configuration parameters

All done mechanically via IDE support.

Release note: None

tbg committed:
server: extract SQLConfig from Config

Release note: None

tbg committed:
server: introduce BothConfig

Not at all in love with this, but it's a step in the right direction.

Release note: None

tbg committed:
base: store spec, not idx, in TempStorageConfig

Release note: None

tbg committed:
server: lift jobAdoptionStopFile declaration out of SQL server

It's declared using the Stores, which we don't want the SQL server to
have to think about.

Release note: None

tbg committed:
server: clean up HistogramWindowInterval

Release note: None

tbg committed:
server: move StorageEngine to BothConfig

SQL needs to know it when creating temp stores.

Release note: None

tbg committed:
server: move DefaultZoneConfig and Locality to BothConfig

SQL needs these for gcjob and the optimizer.

Release note: None

tbg committed:
server: don't pass full Config to SQL Server

Only pass BothConfig union SQLConfig. This is possible now thanks to the
preceding commits.

Release note: None

tbg committed:
server: trim deps in StartTenant

`(*TestServer).StartTenant` now uses only the Gossip instance(*) from
the TestServer and sets everything else up from scratch. This
demonstrates that mod the Gossip dependency, we're en route to
running the SQL server in a standalone process.

Informs #49104.

(*): and the Stopper, but that's fine.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49459)
<!-- Reviewable:end -->

miretskiy reviewed (changes requested):
> Reviewed 4 of 19 files at r1, 3 of 7 files at r2, 1 of 5 files at r4, 5 of 11 files at r6, 1 of 3 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49459)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @nvanbenschoten, and @tbg)

___
*[pkg/server/config.go, line 114 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M7ytOqe2Xj7bXNgGGDg:-M7ytOqe2Xj7bXNgGGDh:bnizoiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e19fc6b5811c3495a8234545096f915f2500bede/pkg/server/config.go#L114)):*
> ```Go
> // around here. Consider moving the whole Config shebang into `base` and
> // unifying.
> type BothConfig struct {
> ```

Not a fan of BothConfig name tbh.

SharedServerConfig? BaseServerConfig?  CommonServerConfig?
___
*[pkg/server/config.go, line 271 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M7ythVq2eIUTf68bf93:-M7ythVq2eIUTf68bf94:b-rshypy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e19fc6b5811c3495a8234545096f915f2500bede/pkg/server/config.go#L271)):*
> ```Go
> type SQLConfig struct {
> 	// LeaseManagerConfig holds configuration values specific to the LeaseManager.
> 	SQLLeaseManagerConfig *base.LeaseManagerConfig
> ```

Should we prefix every member of this struct with SQL?  The struct already called SQLConfig, afterall...


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 19 of 19 files at r1, 7 of 7 files at r2, 3 of 3 files at r3, 5 of 5 files at r4, 2 of 2 files at r5, 11 of 11 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 3 of 3 files at r9, 3 of 3 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-:-M8Ge0ILCwTM6M-vha9v:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @tbg)

___
*[pkg/server/config.go, line 349 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M8GdSsmA9dkl8g1GxU0:-M8GdSsmA9dkl8g1GxU1:b-8avzkd) ([raw file](https://github.com/cockroachdb/cockroach/blob/94984880d2d7b88c41d60a8ac332640c2969b2e9/pkg/server/config.go#L349)):*
> ```Go
> 		MaxOffset:         MaxOffsetType(base.DefaultMaxClockOffset),
> 		StorageEngine:     storage.DefaultStorageEngine,
> 		DefaultZoneConfig: zonepb.DefaultZoneConfig(),
> ```

nit: keep these in the same order as the fields are in.
___
*[pkg/server/config.go, line 355 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M8GdYQdBHgWP4mlGCWE:-M8GdYQdBHgWP4mlGCWF:b-g199fb) ([raw file](https://github.com/cockroachdb/cockroach/blob/94984880d2d7b88c41d60a8ac332640c2969b2e9/pkg/server/config.go#L355)):*
> ```Go
> 		BothConfig:                     bothCfg,
> 		SQLConfig:                      sqlCfg,
> 		DefaultSystemZoneConfig:        zonepb.DefaultSystemZoneConfig(),
> ```

Even if this doesn't strictly need to be in `BothConfig`, it probably makes sense to keep it next to `DefaultZoneConfig`.
___
*[pkg/server/config.go, line 114 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M7ytOqe2Xj7bXNgGGDg:-M8Gc0UW8vo2sojTim-s:b-ipwf69) ([raw file](https://github.com/cockroachdb/cockroach/blob/e19fc6b5811c3495a8234545096f915f2500bede/pkg/server/config.go#L114)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Not a fan of BothConfig name tbh.

SharedServerConfig? BaseServerConfig?  CommonServerConfig?
</blockquote></details>

+1. Maybe just SharedConfig, since this is in the "server" pkg.
___
*[pkg/server/config.go, line 271 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M7ythVq2eIUTf68bf93:-M8GbjxR8iWUzAW6uTCq:b-u2z81u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e19fc6b5811c3495a8234545096f915f2500bede/pkg/server/config.go#L271)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Should we prefix every member of this struct with SQL?  The struct already called SQLConfig, afterall...
</blockquote></details>

I had the same question. I think we can strip the prefix now that this config struct is split.


<!-- Sent from Reviewable.io -->

tbg committed:
server: hard-code cluster name in StartTenant

Release note: None

tbg committed:
server: strip SQL prefix from fields in SQLOptions

During the preceding refactors, the fields were renamed with a SQL
prefix, but now that they are grouped together in SQLOptions, this
is no longer necessary.

Also renames BothConfig -> BaseConfig.

Release note: None

tbg reviewed (commented):
TFTRs! Let's see if this will turn green.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-:-M8Kt7ZtA-2IW2ByN1qu:br3lvuq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @nvanbenschoten)

___
*[pkg/server/config.go, line 355 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M8GdYQdBHgWP4mlGCWE:-M8KsC0u6J57--Ye1edG:b-z2zpll) ([raw file](https://github.com/cockroachdb/cockroach/blob/94984880d2d7b88c41d60a8ac332640c2969b2e9/pkg/server/config.go#L355)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Even if this doesn't strictly need to be in `BothConfig`, it probably makes sense to keep it next to `DefaultZoneConfig`.
</blockquote></details>

Are you suggesting that I move it to `BothConfig`? I'd rather not - the tenants should stop having to know `DefaultZoneConfig` instead. Let's chat over in https://github.com/cockroachdb/cockroach/issues/49445.
___
*[pkg/server/config.go, line 114 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M7ytOqe2Xj7bXNgGGDg:-M8Kqncs3LiJY62Lsr1Q:b-w4cib) ([raw file](https://github.com/cockroachdb/cockroach/blob/e19fc6b5811c3495a8234545096f915f2500bede/pkg/server/config.go#L114)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1. Maybe just SharedConfig, since this is in the "server" pkg.
</blockquote></details>

How about `BaseConfig` (it embeds base.Config)? Happy to use SharedConfig if that seems better.
___
*[pkg/server/config.go, line 271 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M7ythVq2eIUTf68bf93:-M8KrnNs0sR0klqiImb-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e19fc6b5811c3495a8234545096f915f2500bede/pkg/server/config.go#L271)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I had the same question. I think we can strip the prefix now that this config struct is split.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
Reviewable shows few files w/out any changes.  Please double check.

> Reviewed 4 of 11 files at r16, 1 of 3 files at r20, 11 of 21 files at r22.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-:-M8LVS561ppROAjK1vwH:byfrtkc)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @miretskiy, @nvanbenschoten, and @tbg)

___
*[pkg/jobs/registry_external_test.go, line 92 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M8LUmzm9ZMBQ-GkMpGl:-M8LUmzn41ftcSQEauyd:b-ossng6) ([raw file](https://github.com/cockroachdb/cockroach/blob/59326fee94de4a18dd1e605e2e73a5d5ef58b800/pkg/jobs/registry_external_test.go#L92)):*
> ```Go
> 		r := jobs.MakeRegistry(
> 			ac, s.Stopper(), clock, nodeLiveness, db, s.InternalExecutor().(sqlutil.InternalExecutor),
> 			idContainer, s.ClusterSettings(), (*base.Config)(nil).HistogramWindowInterval(), jobs.FakePHS, "",
> ```

I saw this in few files... SHould we have a method for this?
Maybe GetDefaultHistogramWindowInterval()?
___
*[pkg/server/config.go, line 114 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M7ytOqe2Xj7bXNgGGDg:-M8LV24b7M5bAw3lvX2J:bjo7iq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e19fc6b5811c3495a8234545096f915f2500bede/pkg/server/config.go#L114)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

How about `BaseConfig` (it embeds base.Config)? Happy to use SharedConfig if that seems better.
</blockquote></details>

+1 on base config.
___
*[pkg/server/config.go, line 110 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M8LUyxb7feU40F9isXt:-M8LUyxb7feU40F9isXu:b-21twk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/59326fee94de4a18dd1e605e2e73a5d5ef58b800/pkg/server/config.go#L110)):*
> ```Go
> // BaseConfig holds parameters that are needed to setup either a KV or a SQL
> // server.
> type BaseConfig struct {
> ```

i like base config name.
___
*[pkg/server/testserver.go, line 458 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M8LVJs1E5znqWWT3hsf:-M8LVJs1E5znqWWT3hsg:b4xt41s) ([raw file](https://github.com/cockroachdb/cockroach/blob/59326fee94de4a18dd1e605e2e73a5d5ef58b800/pkg/server/testserver.go#L458)):*
> ```Go
> 		cfg := makeTestConfig(st)
> 		sqlConfig = cfg.SQLConfig
> 		bothConfig = cfg.BaseConfig
> ```

do you want to rename bothConfig?


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod @miretskiy's remaining questions.

> Reviewed 6 of 11 files at r16, 1 of 3 files at r20, 21 of 21 files at r22.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-:-M8Mnq2G7Xs8FYP4RtH-:b-jydlbv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @tbg)

___
*[pkg/base/config.go, line 713 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M8Mn7vz18VYjt8JGkDz:-M8Mn7vz18VYjt8JGkE-:b-d9w5hw) ([raw file](https://github.com/cockroachdb/cockroach/blob/59326fee94de4a18dd1e605e2e73a5d5ef58b800/pkg/base/config.go#L713)):*
> ```Go
> // ExternalIODirConfig describes various configuration options pertaining
> // to external storage implementations.
> type ExternalIOConfig struct {
> ```

Did you want to rename this struct? Was this accidental?
___
*[pkg/server/config.go, line 355 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M8GdYQdBHgWP4mlGCWE:-M8MnYBDA5tP3FZe7fwA:b-eqlqw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/94984880d2d7b88c41d60a8ac332640c2969b2e9/pkg/server/config.go#L355)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Are you suggesting that I move it to `BothConfig`? I'd rather not - the tenants should stop having to know `DefaultZoneConfig` instead. Let's chat over in https://github.com/cockroachdb/cockroach/issues/49445.
</blockquote></details>

SGTM if the goal is to eventually move this back.xx
___
*[pkg/server/config.go, line 114 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M7ytOqe2Xj7bXNgGGDg:-M8MnV6k19hBEyZLSKLI:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/e19fc6b5811c3495a8234545096f915f2500bede/pkg/server/config.go#L114)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

+1 on base config.
</blockquote></details>

+1


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> Reviewable shows few files w/out any changes. Please double check.

I think this is because of the SQL prefix that was added in early commits and removed later.

> Dismissed @miretskiy from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-:-M8PTBW077TRmZiaU80E:b89qbpe)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @nvanbenschoten)

___
*[pkg/base/config.go, line 713 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M8Mn7vz18VYjt8JGkDz:-M8PTM7d4YApc4CeQz8w:b-ki6v2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/59326fee94de4a18dd1e605e2e73a5d5ef58b800/pkg/base/config.go#L713)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did you want to rename this struct? Was this accidental?
</blockquote></details>

Definitely accidental. Thanks for catching that.
___
*[pkg/jobs/registry_external_test.go, line 92 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M8LUmzm9ZMBQ-GkMpGl:-M8PU6lr9cli9Az6il6f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59326fee94de4a18dd1e605e2e73a5d5ef58b800/pkg/jobs/registry_external_test.go#L92)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I saw this in few files... SHould we have a method for this?
Maybe GetDefaultHistogramWindowInterval()?
</blockquote></details>

Done.
___
*[pkg/server/testserver.go, line 458 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/49459#-M8LVJs1E5znqWWT3hsf:-M8PVeYCCezTf64hNer4:b-xq3u9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/59326fee94de4a18dd1e605e2e73a5d5ef58b800/pkg/server/testserver.go#L458)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do you want to rename bothConfig?
</blockquote></details>

Had missed that, thanks. Done.


<!-- Sent from Reviewable.io -->

tbg commented:
bors r=nvanbenschoten
